text/3440-virt-self.md

```diff
@@ -0,0 +1,189 @@
+
+What I came up is passing a &dyn System virtualized self so now System looked like this
+
+```rust
+pub trait System {
+    fn op1(&self, vself: &dyn System) -> SystemOp1;
+    fn op2(&self, vself: &dyn System) -> SystemOp2;
+}
+impl System for SystemA {
+    fn op1(&self, vself: &dyn System) -> SystemOp1 {
+        vself.op2() * 5
+    }
+    fn op2(&self, vself: &dyn System) -> SystemOp2 {
+        3
+    }
+}
+```
+
+And then calling `op2` with essentially another copy of `self`:
+```rust
+let system = DoubleOp2<SystemA> {};
+system.op2(&system2); // works!
+```
+
+Of course this design is a bit clunky, but is very powerful. It allows complete control over what is virtualized and what isn't at any point in the call chain for every implementation.
+
+As I have been thinking about this more and more, I wondered what Rust with this in the language would look like. This RFC proposes a sample syntax (bikeshedding welcome, but please focus on the concepts).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Essentially, we now allow `&virt self` as a shorthand for the above.
+
+```rust
+pub trait System {
+    fn op1(&virt self) -> SystemOp1;
+    fn op2(&virt self) -> SystemOp2;
+}
+```
+
+The effect is that rust will manage two pointers - one concretely typed and one dynamic. The The concretely typed (i.e. `self`) follows the existing Rust rules. The dynamic one is either `self` as `dyn System` again or an existing dynamic `self` (this is like choosing `self` or `vself` in the previous section).
+
+Syntax could be e.g.
+```rust
+impl System for SystemA {
+    fn op1(&virt self) -> SystemOp1 {
+        // alternative 1
+        virt(self).op2() * 5
+        // alternative 2, I like this one most because it reminds me of C
+        self->op2() * 5
+        // something else...
+    }
+    fn op2(&virt self) -> SystemOp2 {
+        3
+    }
+}
+```
+
+Working with the second syntax, the difference between `self.op2()` and `self->op2()` would be static vs dynamic dispatch. In other words, it would allow us to call `DoubleOp2<SystemA>::op2()` from within `SystemA::op1()` when the call to it is made from `DoubleOp2<SystemA>::op1()`.
+
+If the called method is also declared virt, using `self->op2()` will retain the `vself` the same as originally passed to `op1`. Otherwise, it will replace it by `self as &dyn System`.
+
+Outside of traits, `a->b()` wouldn't compile. 
+
+```[error-xxx] trait virtual methods can only be called virtually from within the trait```
+
+Instead, only the traditional syntax of `a.b()` will be allowed and this would simply use a `vself` of `a as &dyn Trait`.
+
+I believe this feature will make Rust more user-friendly to people more inclined to think in OOP terms or who (like me) simply found themselves writing code in a domain that is very amenable to an OOP approach.
+
+The syntax changes are relatively minimal and there is no extra cost for code that does not use this feature.
```

Comment:
Care to elaborate here? Rust already knows when you're calling into a trait (forcing you to import it in scope) so any such call should simply use the ref to static as the vself? Or did i miss your point?

Reply:
My statement is that by just adding this feature, now functions that internally reference global state can have dependencies that are hard-to-impossible to resolve at compile time, so they can't be statically resolved and thus devirtualized as easily as they could before. This would impact even programs _that do not use this feature_. It could be that this doesn't actually make anything worse for optimizers, but I suspect it does, so it's probably the case that it is not true that it imposes no costs anywhere else.